src/compiler/scanner.ts

```diff
@@ -1646,16 +1646,28 @@ export function createScanner(languageVersion: ScriptTarget,
             if (pos >= end) {
                 return token = SyntaxKind.EndOfFileToken;
             }
-            const ch = codePointAt(text, pos);
 
-            // Special handling for shebang
-            if (ch === CharacterCodes.hash && pos === 0 && isShebangTrivia(text, pos)) {
-                pos = scanShebangTrivia(text, pos);
-                if (skipTrivia) {
-                    continue;
+            const ch = codePointAt(text, pos);

```

Comment:
We handle some of the file encoding at https://github.com/microsoft/TypeScript/blob/main/src/compiler/sys.ts#L1784 Instead of scanner, would we want to move this there and throw with error that gets reported as part of reading file itself? With your approach the file is read and kept in memory but with sys approach though it wont be in memory, external toolings need to handle these kind of things on their own which could probably be ok?

Reply:
I originally had changed this in `sys` but it's very awkward to thread up from here, since the only thing `readFile` can do is return `undefined` (or maybe lie about the file contents); the former case we'd report a confusing "file X not found", and I wasn't sure what we should even do in the latter case (return empty string maybe?). I also figured the scanner would be a better place since there might be alternate `sys` implementations floating around and we never ever want to parse a file with a `u+fffd` in the leading character position.

What are you thinking `readFile` should do more specifically?

Reply:
If we throw from there with the error message that you want to be reported, the error should be populated as part of building?